Control

zmerino opened the pull request:
Starting the review of code updates Madison has made for the control/starkshift modules. I'll review this closely sometime between now and Thursday.

zmerino commented on unit_tests/test_control.py:
```diff
@@ -0,0 +1,216 @@
+from random import random
+
+from qudipy.control import shapes
+from qudipy.starkshift import starkshit
```
Frustrations with the starkshift module I see :joy:

zmerino commented:
@madisonXschuetze At a glance, things are looking good! I just have two initial general comments before I look over everything more thoroughly:

1) Can you modify the two unit test you wrote to resemble unit tests which utilize the pytest package. This is a popular package used for testing that has many useful features (like fixtures) and helps reduce the amount of code needed for the test scripts. The most recent merge with master now contains a tutorial discussing some aspects of unit testing with pytest and has an example for the operator module. Note, that there is now a nested directory structure for the unit test that will enable several things, mainly, being able to run unit tests with different module scopes as opposed to having to test all of qudipy at once.

2) Can you merge master into your branch and push the updated Control branch to github. This will help reviewing the relevant changes.

Thanks!

madisonXschuetze commented:
@zmerino I looked at pytest documentation and your unit tests for the operator module and it looks very neat! I will rework the test_control module and the test_starkshift module to use pytest instead. 

After I finish reworking this code, I will merge the master branch into my branch and push this again!

zmerino commented:
I'll start review some things about this pull request over today and tomorrow. A couple of small things:

Can you remove this "unit test" directory and the unit test file inside of it. This unit test will be added later on once the nextnano_conversion PR is approved.
https://github.com/mainCSG/QuDiPy/blob/c40f439fd56e4094115912d18f69e207090ce856/unit%20tests/test_process_nextnano_files.py#L2

A minor change, but can you structure the two unit tests that you've written i.e. `test_control.py` and `test_starkshift.py` to be modelled after the file structure of the code in the qutils_test directory. For a single unit test, this is not useful, but as we add test in future having this file structure will prove useful.
https://github.com/mainCSG/QuDiPy/blob/c40f439fd56e4094115912d18f69e207090ce856/unit_tests/qutils_test/conftest.py#L26-L28

zmerino commented:
@madisonXschuetze I was just looking over some of the code for the control branch and noticed that it is still not updated with master. For example, in the Control branch the const_dict function for the spinsimulator module was:

https://github.com/mainCSG/QuDiPy/blob/509485be924c88cba43f3049a4f439580c1770a9/qudipy/spinsimulator/spin_simulator.py#L242-L244

And under the master branch it is:

https://github.com/mainCSG/QuDiPy/blob/b75bb325477f8238580245c56a8cc373ecb8d70b/qudipy/spinsimulator/spin_simulator.py#L247-L249

If you haven't checked it out already, I certainly recommend using git desktop to make merging one branch into another quite seamless. Let me know if you have any difficulties and I'm happy to help merge master into the Control branch.

zmerino commented:
Looks like the object tag `self.` was left behind from when the unittest package was in use. Some function inputs will need to be added most likely. The preprocess_nextnano_test/conftest.py file might be of interest here, but may not necessarily need to be used. 

https://github.com/mainCSG/QuDiPy/blob/892c526f752ca7172d40cfe8dbcf18b915d0ebc3/unit_tests/starkshift_test/test_starkshift.py#L24-L25

zmerino commented on unit_tests/control_test/test_control.py:
```diff
@@ -0,0 +1,217 @@
+	return [square_shape, triangle_shape, wide_gauss, narrow_gauss]
+
+@pytest.fixture(scope="module")
+def g_interp(v):
```
Does there need to be another fixture defined for "v" or should this be some global variable?

madisonXschuetze replied:
It would be a globally defined variable - the rot() function requires a callable function as a parameter, this fixture is that function. This callable function needs to take Anyof(Float, Int, tuple, array, List).

zmerino commented on unit_tests/control_test/test_control.py:
```diff
@@ -0,0 +1,217 @@
+
+
+## *** run tests ***
+def test_arbitrary_rot(shapes, g_interp, rot_save):
```
fixture defined above is shapes_list

madisonXschuetze replied:
I've fixed this in my code! Thank you for pointing it out :)

zmerino commented on unit_tests/control_test/test_control.py:
```diff
@@ -0,0 +1,217 @@
+
+## *** run tests ***
+def test_arbitrary_rot(shapes, g_interp, rot_save):
+	B0 = balance_zeeman(g_interp, square_shape, 1000)
```
Later on B_0 is used inplace of B0

madisonXschuetze replied:
Fixed :)

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+    "import copy\n",
+    "from scipy.interpolate import interp1d\n",
+    "\n",
+    "#potential module\n",
```
Very minor, but let's change the comments to have a leading space after # and to make sure the comment is capitalized. So in this case `# Potential`

madisonXschuetze replied:
Fixed!

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+    "#potential module\n",
+    "import qudipy.potential as pot\n",
+    "\n",
+    "# for solving Schrod eq-n and Stark shift stuff\n",
```
`# For`

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+    "import qudipy.qutils as qt\n",
+    "import qudipy.starkshift as ss\n",
+    "\n",
+    "#Spin simulator module\n",
```
`# Spin`

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+    "\n",
+    "#Spin simulator module\n",
+    "import qudipy.spinsimulator.spin_simulator as sps\n",
+    "#Constant pulse generating module\n",
```
`# Constant`

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+    "#Constant pulse generating module\n",
+    "import qudipy.spinsimulator.pulse_generators as pulsegens\n",
+    "\n",
+    "#Circuit module containing control pulses and ideal circuits\n",
```
`# Circuit`

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+    "#Circuit module containing control pulses and ideal circuits\n",
+    "import qudipy.circuit as circ\n",
+    "\n",
+    "#Control shapes and voltage pulse routines\n",
```
`# Control`

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+    "#Control shapes and voltage pulse routines\n",
+    "from qudipy.control import shapes, voltage_pulses, balance_zeeman\n",
+    "\n",
+    "#File Parsers to load data\n",
```
`# File` and etc

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
```
Similar to other tutorials, let's add a small description describing what the tutorial aims to accomplish

madisonXschuetze replied:
Sounds good ! I have added a description to the best of my ability but I think Bohdan will have a more appropriate description.

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+   "outputs": [],
+   "source": [
+    "V_0 = 0.65  # pulse amplitude, V\n",
+    "dV = 0.5    # offset, V\n",
```
Can units be added in a comment for constants as well as a small description of any variables/constants that are no super self explanatory

zmerino replied:
Or it might even be better to split this cell up into two parts, i.e. the constant definitions and the pulse creation, then have two markdown cells describing each code block with a brief description?

madisonXschuetze replied:
Sure! I can do this :) I will make it more clear

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "input_nextnano = os.path.join(sys.path[0],'QuDiPy tutorial data','Nextnano simulations','1_dot_P_0p5_0p7')\n",
```
A comment as a reminder to myself, but the 1_dot_P_0p5_0p7 data files should be parsed down to only the essentials for the tutorial purposes. Currently, the directory is ~2GB.

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
FYI, I changed this code block, but want to add a more detailed comment/markdown message

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Constructing a z-slice for the data"
```
Can we make this markdown statement a little more descriptive, something like, "Constructing the 2D slice along the z-axis for the 3D potential data which was imported"

zmerino replied:
I think this comment applies generally for this tutorial (as well as other already created tutorials)

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+   "source": [
+    "\n",
+    "# Now we define the field types we wish to write 2D slices for. Either potential or the electric field.\n",
+    "pot.process_nextnano.write_data(input_nextnano,output_preprocessed, slice=z, f_type=['potential','field'])\n"
```
Let's add `save_flag = pot.process_nextnano.write_data...` to suppress the returned bool value from being displayed

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "loaded_data_pot = pot.load_potentials(ctrl_vals, ctrl_names,\n",
```
Along with the TODO statement can we plot the potential and electric field 2D slices in a subplot like previously done in the tutorial

zmerino commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+   "metadata": {},
+   "source": [
+    "## True g(V)\n",
+    "Now we create a ''true'' g-factor dependency on voltage.\n",
```
Let's describe what the "true" g-factor is in reference too

zmerino commented on qudipy/control/shapes.py:
```diff
@@ -0,0 +1,171 @@
+'''
+    Module with a library of functions that define pulse shapes. 
+    Each of them returns a shape function that gives the pulse value(s)
+    when applied on the desired value(s) of time.
```
Go ahead and add your author ship here `@author :`. We want you to get credit where credit is due :slightly_smiling_face:

zmerino commented on qudipy/control/shapes.py:
```diff
@@ -0,0 +1,171 @@
+def square(t_start=0, t_end=1, amp=1, offset=0):
+    '''
+        Creates a square pulse of amplitude amp and that continues
+        from t_start until t_end. The offset value before and after the pulse 
```
` The voltage offset value`

zmerino commented on qudipy/control/shapes.py:
```diff
@@ -0,0 +1,171 @@
+        '''
+    def square_pulse(t):    
+        # making the time input a float array
+        if isinstance(t, (int,float,list,tuple,range)):
```
Nice call on the data type check, but can you convert the if statement to a try/exception clause. Something like,

```
try:
     t = np.array(t, dtype=float)
except ValueError:
     print(f'Unable to convert {dtype(t)} object to float array')
```
Otherwise the exception will occur when trying to use the `t` later on.

madisonXschuetze replied:
Fixed :)

zmerino commented on qudipy/control/shapes.py:
```diff
@@ -0,0 +1,171 @@
+        '''
+    def triangular_pulse(t):    
+        # making the time input a float array
+        if isinstance(t, (int,float,list,tuple,range)):
```
Same comment as above for the square pulse

zmerino commented on qudipy/control/shapes.py:
```diff
@@ -0,0 +1,171 @@
+    def shifted_gauss_pulse(t):
+
+        # making the time input a float array
+        if isinstance(t, (int,float,list,tuple,range)):
```
Same comment as above for the square pulse

madisonXschuetze replied:
Fixed :)

zmerino commented on qudipy/control/shapes.py:
```diff
@@ -0,0 +1,171 @@
+        return pulse_arr
+    return shifted_gauss_pulse
+
+if __name__=='__main__':
```
Let's remove this once testing/reviewing is done and we're ready to merge the PR

madisonXschuetze commented on tutorials/test_single_qubit_gates.ipynb:
```diff
@@ -0,0 +1,659 @@
+    "\n",
+    "#Constants class\n",
+    "from qudipy.utils.constants import Constants\n",
+    "from scipy.constants import e,m_e\n",
```
I have taken it out :)

hromecB commented on qudipy/chargestability/csd_analysis.py:
```diff
@@ -475,12 +476,30 @@ def __plot_heatmap(self, data, x_values, y_values, x_label, y_label, cbar=True,
+        cmap = mpl.cm.magma
+        
+        # Create Plot
+        if x_values is not None and x_values is not None:
```
Repetitive statement. Did you mean `if x_values is not None and y_values is not None:` ?

hromecB commented on qudipy/control/voltage_pulses.py:
```diff
@@ -0,0 +1,570 @@
+def rot(rotation_axis, theta, n_qubits, active_qubits, 
+                delta_g_interp, v_unit_shape, B_0, num_val=100):
+    '''
+    @author: Madi & Others (TODO specify others)
```
I think you can move this statement to the beginning of the file

hromecB commented on qudipy/control/voltage_pulses.py:
```diff
@@ -0,0 +1,570 @@
+        Positions of qubits (starting from 1, not 0!) on which the pulse
+         is supposed to act. The rest will be made inactive.
+    delta_g_interp: function
+        (ideally, but only one interpolating function for now)
```
There's something wrong with this sentence. 
I'll need to recollect what I meant by "ideally, but" 
Just add a TODO here

hromecB commented on qudipy/control/voltage_pulses.py:
```diff
@@ -0,0 +1,570 @@
+    else:
+        # *** ARBITRARY ROTATION ***
+        #create variables for Vector Components to speed up run-time
+        n_x = rotation_axis[0]
+        n_y = rotation_axis[1]
+        n_z = rotation_axis[2]
```
It's the same thing but with a more concise syntax
```suggestion
        n_x, n_y, n_z = rotation_axis
```

hromecB commented on qudipy/control/voltage_pulses.py:
```diff
@@ -0,0 +1,570 @@
+                return delta_g_diff / param
+    
+        # Find Shape
+        if n_z == 1:
+            S = find_shape(active = True) #get shape
+        else:
+            S = find_shape(active = False) #get shape
```
Can be replaced with a single line: 
```suggestion
            S = find_shape(active = (n_z == 1)) #get shape
```

hromecB commented on qudipy/control/voltage_pulses.py:
```diff
@@ -0,0 +1,570 @@
+
+            Returns
+            ---------------
+            List 
```
Actually, 1D array, not list

hromecB commented on qudipy/control/voltage_pulses.py:
```diff
@@ -0,0 +1,570 @@
+    #TODO pandas recall how to add values
+    
+
+def create_ctlp_file(CntrlPulse, read = False, get_dir = False):
```
This function should be moved to another module, it's not directly related to the design of voltage pulses. 
I recommend that you create a separate file in the Control Pulse folder

hromecB commented:
@zmerino 
# PULL REQUEST READY FOR FINAL REVIEW

Highlights: 
- State class created instead of SpinSys class. Now DotArray and State classes are parts of `qudipy.system` module
- Designer and DesignerEffective classes created for CSP pulse design in terms of effective parameters only (for now)
- ControlPulse plotting routine improved
- Code for all qubit rotations, SWAP^k, and CPHASE(α) gates is re-written in terms of shape functions & constant vectors
- Effective pulse design and qubit evolution are showcased in `Effective_spin_control.ipynb` tutorial that replaces the old Spin Simulator, Test_single_qubit_gates, and Single_gates tutorials
- new logic for universal and material constants: 
    - universal constants are just parts of `qudipy.utils.constants` module: no need to create Constant class objects anymore
    - Material dataclass handles all material-related parameters. The library of standard Material objects is pre-calculated, stored in  `qudipy.utils.constants` module, and can be accessed by the `select_material` function from there. This allows for a user to define new or customize known materials
 - functions begin to get type annotations ( including custom type definitions) for more intuitive coding experience

## (Non-critical) features that will likely be added during review, before the pull request is merged
- Adding a section to the new tutorial with dissipative evolution (Rabi or exchange oscillations)
- Transforming dictionary with qubit evolution data to EvolutionData class with:
    - information about all pulses that drove the evolution added
    - custom `plot` method for quick plotting by a user in a tutorial, instead of long lines of Matplotlib figure/axis definitions

zmerino commented:
Changed:
https://github.com/mainCSG/QuDiPy/blob/2e8005798b991bbddfb2bf73eb1c6afaebcfb987/qudipy/shuttling/real_space.py#L318
To: `f_path = os.path.join(save_dir, save_name)`

And updated "Electron Shuttling.ipynb" tutorial accordingly

zmerino commented:
Changed:
https://github.com/mainCSG/QuDiPy/blob/2e8005798b991bbddfb2bf73eb1c6afaebcfb987/qudipy/potential/create_potential.py#L141-L143
To:
```
			mu =  10 * v_i * constants.e
			omega = self.omegas[i]
			well = 1/2 * omega * constants.e *\
```

For "Loading Potentials.ipynb". Also, corrected me -> m0 where necessary

zmerino commented:
FYI, The last cell "2.3 Plot Fitted Potential" in the "Fittining.ipynb" tutorial produces NaNs which are not plotted.

zmerino commented:
Updated "Operator Creation.ipynb" tutorial.

zmerino commented:
FYI, "Manipulating .ctrlp and .qcirc files.ipynb" is missing "/QuDiPy tutorial data/Control pulses/ROTX_1_2.ctrlp" data.

To be added back into the tutorial data set, or modified so that it is first generated.

zmerino commented:
Everything else, looks good enough to merge to master.
